### ISSUE TITLE
New credentials validation mechanism - False positives due to too long waiting time or parallel search for remote 

### ISSUE THREAD

**Issue by FALLAI-Denis** [IBM/zopeneditor-about#197]
## Development environment used

- Z Open Editor version: 1.4.1 (with fix for #142)
- Editor Platform
  - [X] Visual Studio Code
  - [ ] Red Hat CodeReady Workspaces
  - [ ] Eclipse Che
  - [ ] Standalone Theia
- Editor Platform Version: 1.62.3
- Operating System: Windows 10 "1909"
- Java Version: java version "1.8.0_102" / Java(TM) SE Runtime Environment (build 1.8.0_102-b14) / Java HotSpot(TM) 64-Bit Server VM (build 25.102-b14, mixed mode)
- [ ] Related to RSE API?
  - RSE API Plugin version:
  - Zowe CLI version:
  - Node.js version:
- Logs attached: yes 
[Code_logs_20211126T114933_exthost1_IBM.zopeneditor_zopen.txt](https://github.com/IBM/zopeneditor-about/files/7608951/Code_logs_20211126T114933_exthost1_IBM.zopeneditor_zopen.txt)

## Problem Description

Detailed steps for reproducing the problem:

After starting VS Code, opening 3 COBOL source files of approximately 3000 lines each, and using many and same COPYBOOKs.

Error message on credentials when in reality they are valid.

Partial loading of COPYBOOKs: the first references are not resolved, the following references are.

### Observed behavior

A message indicating that the credentials are invalid is raised:
![image](https://user-images.githubusercontent.com/63861965/143587407-5ec18888-22da-4f77-abe0-4b37fca9b801.png)

When analyzing the log, we see:
- Z Open Editor launches a request to validate the credentials (read the SYS1.PROCLIB file):
`2021-11-26T12:41:45.995+0100 DEBUG: ZoweResourceResolver.checkMVSConnection: Running a Zowe profile connection test using the profile "GMVS" by reading data set "sys1.proclib".
`
- Z Open Editor launches many requests to search for a COPYBOOK on mvs remote system:
`2021-11-26T12:41:45.565+0100 DEBUG: ZLanguagesResourceResolver.documentRequested: The COBOL language server requested the following include file: "AAAAVARC". The ID for this request was assigned as "atba2e" There are 1 pending requests being processed.`
`2021-11-26T12:41:46.034+0100 DEBUG: ZLanguagesResourceResolver.documentRequested: The COBOL language server requested the following include file: "AAAAVARC". The ID for this request was assigned as "qvjps" There are 2 pending requests being processed.`
`2021-11-26T12:41:48.661+0100 DEBUG: ZLanguagesResourceResolver.documentRequested: The COBOL language server requested the following include file: "AAAAVARC". The ID for this request was assigned as "x6sld" There are 3 pending requests being processed.`

At the same time, many (7?) user stc "IZUFPROC" have been started on mvs.

First user stc:
```
[12:41:56.381967] ServletDispatcher[I]: Received request: GET /ds?dslevel=sys1.proclib                                  
[12:41:56.383226] ServletDispatcher[I]: z/OSMF transaction: tx0000000000008CE6                                          
[12:41:56.383261] ServletDispatcher[I]: Invoking Servlet: SELECT  PGM(IZUGLSDS) PARM(-LI,t -q 5439498 -r 0x298e2028 )   
[12:41:56.421506] ServletDispatcher[I]: Servlet IZUGLSDS completed,  RC=0                                               
[12:41:57.277449] ServletDispatcher[I]: Received request: GET /ds/DSGETU.PERMIBL.SIRIS.COPY(AAAAVARC)                      
[12:41:57.277478] ServletDispatcher[I]: z/OSMF transaction: tx0000000000008CE8                                             
[12:41:57.277515] ServletDispatcher[I]: Invoking Servlet: SELECT  PGM(IZUGGTDS) PARM(-LI,t -q 5439498 -r 0x298e2688 )      
[12:41:57.309773] ServletDispatcher[I]: Servlet IZUGGTDS completed,  RC=0      
[12:42:04.630386] ServletDispatcher[I]: Received request: GET /ds/DSGETU.PERMIBL.SIRIS.COPY(AAADACCE)                    
[12:42:04.630421] ServletDispatcher[I]: z/OSMF transaction: tx0000000000008CF3                                           
[12:42:04.630458] ServletDispatcher[I]: Invoking Servlet: SELECT  PGM(IZUGGTDS) PARM(-LI,t -q 5439498 -r 0x298e2688 )    
[12:42:04.784695] ServletDispatcher[I]: Servlet IZUGGTDS completed,  RC=0                                                                                            
...
```

2nd user stc:
```
[12:41:58.359970] ServletDispatcher[I]: Received request: GET /ds/DSGETU.PERMIBL.SIRIS.COPY(AAADPAC0)                       
[12:41:58.361427] ServletDispatcher[I]: z/OSMF transaction: tx0000000000008CEA                                              
[12:41:58.361506] ServletDispatcher[I]: Invoking Servlet: SELECT  PGM(IZUGGTDS) PARM(-LI,t -q 7208971 -r 0x298e2028 )       
[12:41:58.392528] ServletDispatcher[I]: Servlet IZUGGTDS completed,  RC=0                                                   
[12:41:58.509948] ServletDispatcher[I]: Received request: GET /ds/DSGETU.PERMIBL.SIRIS.COPY(AAADPAC0)                       
[12:41:58.509960] ServletDispatcher[I]: z/OSMF transaction: tx0000000000008CEE                                              
[12:41:58.509986] ServletDispatcher[I]: Invoking Servlet: SELECT  PGM(IZUGGTDS) PARM(-LI,t -q 7208971 -r 0x298e2688 )       
[12:41:58.558443] ServletDispatcher[I]: Servlet IZUGGTDS completed,  RC=0   
[12:42:03.10311] ServletDispatcher[I]: Received request: GET /ds/DSGETU.PERMIBL.SIRIS.COPY(AAADACCE)                          
[12:42:03.10349] ServletDispatcher[I]: z/OSMF transaction: tx0000000000008CF0                                                 
[12:42:03.10384] ServletDispatcher[I]: Invoking Servlet: SELECT  PGM(IZUGGTDS) PARM(-LI,t -q 7208971 -r 0x298e2688 )          
[12:42:03.558695] ServletDispatcher[I]: Servlet IZUGGTDS completed,  RC=0                                                     
...
```

3rd user stc:
```
[12:41:59.517929] ServletDispatcher[I]: Received request: GET /ds/DSGETU.PERMIBL.SIRIS.COPY(AAADPAC0)                         
[12:41:59.519272] ServletDispatcher[I]: z/OSMF transaction: tx0000000000008CEC                                                
[12:41:59.519312] ServletDispatcher[I]: Invoking Servlet: SELECT  PGM(IZUGGTDS) PARM(-LI,t -q 7864334 -r 0x298e2028 )         
[12:41:59.754996] ServletDispatcher[I]: Servlet IZUGGTDS completed,  RC=0           
...
```

These requests seem to take a long time to be processed for Z Open Editor, while at mvs level they seem to have been processed quickly.

After a while, Z Open Editor declares that the COPYBOOKs are not found : `Stopping remote file resolving due to invalid credentials for the Zowe CLI profile`.

### Expected behavior

If the processing time for the credentials validation request is too long, and COPYBOOKs search requests on mvs remote system are pending, a pop-up window should be displayed to indicate it to the user, and ask him if he wants to continue to wait, or if he wants to abandon the search for COPYBOOKs on the remote mvs system (and repeat this as many times as necessary if the user decides to wait).

**Comment by kkrein** [comment]
Ich bin derzeit nicht im Büro.
Während meiner Abwesenheit bitte bei Postkorb SCM-Admin melden!

Die Rückmeldung bezieht sich auf ein Mail mit folgendem Thema:
[IBM/zopeneditor-about] New credentials validation mechanism - False
positives due to too long waiting time (Issue #197)

____________________________________________________________________________________________
Gesendet (c) GRZ/RSG Linz 2021 Agent 'Abwesenheit'

----------------------------------------
Raiffeisen Software GmbH
Firmenbuchnummer: 86804d, Landesgericht Linz
UID: ATU 36961105

Der Austausch von Nachrichten mit dem oben angeführten Absender via E-Mail
dient ausschließlich Informationszwecken. Rechtsgeschäftliche Erklärungen
dürfen über dieses Medium nicht ausgetauscht werden.
Correspondence with above mentioned sender via e-mail is only for
information purposes. This medium is not to be used for the exchange of
legally-binding communications.
----------------------------------------

**Comment by phaumer** [comment]
Thanks. Can you show the error messages in the Z Open Editor log for the failed requests?

**Comment by kkrein** [comment]
Ich bin derzeit nicht im Büro.
Während meiner Abwesenheit bitte bei Postkorb SCM-Admin melden!

Die Rückmeldung bezieht sich auf ein Mail mit folgendem Thema:
Re: [IBM/zopeneditor-about] New credentials validation mechanism - False
positives due to too long waiting time or parallel search for remote
(Issue #197)

____________________________________________________________________________________________
Gesendet (c) GRZ/RSG Linz 2021 Agent 'Abwesenheit'

----------------------------------------
Raiffeisen Software GmbH
Firmenbuchnummer: 86804d, Landesgericht Linz
UID: ATU 36961105

Der Austausch von Nachrichten mit dem oben angeführten Absender via E-Mail
dient ausschließlich Informationszwecken. Rechtsgeschäftliche Erklärungen
dürfen über dieses Medium nicht ausgetauscht werden.
Correspondence with above mentioned sender via e-mail is only for
information purposes. This medium is not to be used for the exchange of
legally-binding communications.
----------------------------------------

**Comment by FALLAI-Denis** [comment]
Hi @phaumer 

I don't know if I understood the request correctly.

The full log was provided with the issue.

I have filtered the searches for COPYBOOKs below.
[Code_logs_20211126T114933_exthost1_IBM.zopeneditor_zopen (filtered).txt](https://github.com/IBM/zopeneditor-about/files/7616717/Code_logs_20211126T114933_exthost1_IBM.zopeneditor_zopen.filtered.txt)

We see that the first request concerns the COPYBOOK "AAAAVARC", with 3 requests in parallel because 3 source files opened, (these files are copies of each other).
- the first request "atba2e" finds the COPYBOOK "AAAAVARC", but is the last to be terminated,
- the second request "qvjps" does not find the COPYBOOK "AAAAVARC", and ends before the first request,
- the third request "x6sld" does not find the COPYBOOK "AAAAVARC", and ends before the first request.

The validation of the Zowe CLI profile was launched after the first request for COPYBOOK "AAAAVARC".

We also see in the log http 404 errors on certain accesses which should not even be triggered (search in a second PDS mvs while the COPYBOOK exists in a first PDS mvs).

It seems to me that there is a synchronization problem between the Zowe CLI profile validation request and the COPYBOOKs search requests on the mvs server triggered in parallel with the validation of the Zowe CLI profile.

**Comment by phaumer** [comment]
My apologies @FALLAI-Denis. I completely missed that you had attached a log file. We will investigate.

**Comment by phaumer** [comment]
Fixed in Z Open Editor v2.0.0. Let us know if this improves the cases you observed.
